sentry_sdk/integrations/gql.py

```diff
@@ -0,0 +1,90 @@
+import gql  # type: ignore[import]
+from graphql import print_ast, get_operation_ast, DocumentNode, VariableDefinitionNode  # type: ignore[import]
+from gql.transport import Transport, AsyncTransport  # type: ignore[import]
+from gql.transport.exceptions import TransportQueryError  # type: ignore[import]
+from sentry_sdk.utils import event_from_exception
+from sentry_sdk.hub import Hub
+from sentry_sdk.integrations import Integration
+from typing import Any, Dict, Optional, Tuple, Union
+
+EventDataType = Dict[str, Union[str, Tuple[VariableDefinitionNode, ...]]]
+
+
+class GQLIntegration(Integration):
+    identifier = "gql"
+
+    @staticmethod
+    def setup_once() -> None:
+        _patch_execute()
+
+
+def _data_from_document(document: DocumentNode) -> EventDataType:
+    operation_ast = get_operation_ast(document)
+    data: EventDataType = {"query": print_ast(document)}
+
+    if operation_ast is not None:
+        data["variables"] = operation_ast.variable_definitions
+        if operation_ast.name is not None:
+            data["operationName"] = operation_ast.name.value
+
+    return data
+
+
+def _transport_method(transport: Union[Transport, AsyncTransport]) -> str:
+    """
+    The RequestsHTTPTransport allows defining the HTTP method; all
+    other transports use POST.
+    """
+    try:
+        return transport.method
+    except AttributeError:
+        return "POST"
+
+
+def _request_info_from_transport(
+    transport: Optional[Union[Transport, AsyncTransport]]
+) -> Dict[str, str]:
+    if transport is None:
+        return {}
+
+    request_info = {
+        "method": _transport_method(transport),
+    }
+
+    try:
+        request_info["url"] = transport.url
+    except AttributeError:
+        pass
+
+    return request_info
+
+
+def _patch_execute() -> None:
+    real_execute = gql.Client.execute
+
+    def sentry_patched_execute(
+        self: gql.Client, document: DocumentNode, *args: Any, **kwargs: Any
+    ) -> Any:
+        try:
+            return real_execute(self, document, *args, **kwargs)
+        except TransportQueryError as e:
+            hub = Hub.current
+            event, hint = event_from_exception(
+                e,
+                client_options=hub.client.options if hub.client is not None else None,
+                mechanism={"type": "gql", "handled": False},
+            )
+            event["request"] = {
+                "data": _data_from_document(document),
+                "api_target": "graphql",
+                **_request_info_from_transport(self.transport),
+            }
+
+            event["contexts"] = {
+                "response": {"data": {"errors": e.errors}, "type": "response"}
+            }
```

Comment:
Couple things here.

1) This is a `EventProcessor` usecase. You define one and add it on the scope and then any event that is reported when that scope is active gets this extra info. Look e.g. [here](https://github.com/getsentry/sentry-python/blob/0dd7d5ff91c99d9a5414d6c55fe6041e28bde130/sentry_sdk/integrations/flask.py#L175) for inspiration, or search for `EventProcessor`.

2) We should also make adding the request/response dependent on whether [send_default_pii](https://github.com/getsentry/sentry-python/blob/0dd7d5ff91c99d9a5414d6c55fe6041e28bde130/sentry_sdk/hub.py#L88) is on to avoid capturing potentially sensitive data. Please also add tests for this.

3) We should make sure we're not potentially rewriting stuff on the event if there's something in `event["request"]` or `event["contexts"]` already. I'm not sure if there can be anything set at this point already, but if we move this to an event processor, then it'll be an issue since other event processors will be setting things on the event.

4) We're constructing the `response` here ourselves, which means it might be different from what is actually returned by the server. Do we have a way of capturing the actual response? E.g. somehow inspecting what `real_execute` returns?

5) Same for the request, we're constructing it from the parsed query -- is there a way to get the actual payload as it'll be sent?

Reply:
1. Done
2. Done
3. I am preserving as much of the `event["request"]` and `event["context"]` as possible, but if there is anything in the specific fields that I am trying to set (e.g. if the request already has a different URL set), this will still get overwritten. Is this okay, or should I rewrite so that we never overwrite anything (which would probably necessarily mean that we lose information captured by the GQL integration)?
4. Getting this information would probably be quite complicated and would probably require significantly changing the code I have written so far due to the way GQL works. This is because the actual HTTP request is made from the GQL transport, of which there are [several different types](https://gql.readthedocs.io/en/stable/transports/index.html) (including some that use WebSockets and other non-HTTP technologies), instead of being made directly from the `Client.execute` function that we are patching here. So, we would likely need to write separate integrations for each GQL transport. And, I think what we are currently capturing should already be sufficient for most use cases, since we are displaying to users the GraphQL query and response, as well as the URL. <br>
Also, since `real_execute` raises the exception, there is no return value for us to be able to inspect here. We have to extract the information we can from the parameters that are passed to `real_execute` and from the error it raises.

5. Same as #&#8203;4 above

Reply:
3. Overwriting single fields with more specific info is fine -- I was more worried about potentially rewriting unrelated fields. E.g. with
```python
            event["contexts"] = {
                "response": {"data": {"errors": e.errors}, "type": "response"}
            }
```
as it was previously, we'd erase any other contexts already set in `event["contexts"]`. But this should not be the case anymore.

4+5. Ideally we would capture the request and response as they were sent and received by the client, but understood. As long as our reconstructed requests/responses are  equivalent this is fine.